src/internal/storage/chunk/task_chain_test.go

```diff
@@ -0,0 +1,83 @@
+package chunk
+
+import (
+	"context"
+	"errors"
+	"math/rand"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/require"
+	"golang.org/x/sync/semaphore"
+)
+
+func TestTaskChain(t *testing.T) {
+	ctx := context.Background()
+	t.Run("Basic", func(t *testing.T) {
+		t.Parallel()
+		numTasks := 5
+		tc := NewTaskChain(ctx, semaphore.NewWeighted(int64(numTasks)))
+		doneParallel := make(chan int, numTasks)
+		doneSerial := make(chan int, numTasks)
+		for i := 0; i < numTasks; i++ {
+			i := i
+			require.NoError(t, tc.CreateTask(func(_ context.Context) (func() error, error) {
+				time.Sleep(time.Duration(numTasks-i-1) * time.Second)
```

Comment:
maybe make the sleep random?

Reply:
The sleep is intentionally longer for earlier tasks to ensure they ran in parallel. So, the order of sends to the done parallel channel should be the reverse of the task creation.